src/main/java/picard/sam/AbstractAlignmentMerger.java

```diff
@@ -751,54 +765,103 @@ private void transferAlignmentInfoToPairedRead(final SAMRecord firstUnaligned, f
+        final int oldStart = rec.getAlignmentStart();
+        final Cigar newCigar = new Cigar();
+        List<CigarElement> cigarElements = new ArrayList<>(oldCigar.getCigarElements()); //need to be modifiable
+        int posShift = 0;
+        boolean foundNonClip = false;
+        for (final CigarElement cigarElement : cigarElements) {
+            final CigarOperator op = cigarElement.getOperator();
+
+            if (op == CigarOperator.SOFT_CLIP) {
+                newCigar.add(new CigarElement(cigarElement.getLength(), CigarOperator.MATCH_OR_MISMATCH));
+                if (!foundNonClip) {
+                    posShift += cigarElement.getLength();
+                }
+            } else {
+                if (!op.isClipping()) {
+                    foundNonClip = true;
+                }
+                newCigar.add(new CigarElement(cigarElement.getLength(), op));
             }
         }
 
-        return clipped;
+        rec.setAlignmentStart(rec.getAlignmentStart() - posShift);
+        rec.setCigar(newCigar);
+        readPosition = SAMRecord.getReadPositionAtReferencePosition(rec, pos, false);
+        rec.setCigar(oldCigar);
+        rec.setAlignmentStart(oldStart);
+        return readPosition;
+    }
+
+    private static void clipRead(final SAMRecord rec, final int clipFrom, final boolean useHardClipping) {
+        // If we are using hard clips, add bases and qualities to SAM tag.
+        if (useHardClipping) {
+            moveClippedBasesToTag(rec, clipFrom);
+        }
+
+        // Actually clip the read
+        CigarUtil.clip3PrimeEndOfRead(rec, clipFrom, useHardClipping ? CigarOperator.HARD_CLIP : CigarOperator.SOFT_CLIP);
+    }
+
+    private static void moveClippedBasesToTag(final SAMRecord rec, final int clipFrom) {
+        if (rec.getAttribute(HARD_CLIPPED_BASES_TAG) != null || rec.getAttribute(HARD_CLIPPED_BASE_QUALITIES_TAG) != null) {
+            throw new PicardException("Record already contains tags for restoring hard clipped bases.  This operation will permanently erase information if it proceeds.");
+        }
+
+        final byte[] bases = rec.getReadBases();
+        final byte[] baseQualities = rec.getBaseQualities();
+        final int readLength = rec.getReadLength();
+
+        final int clipPositionFrom, clipPositionTo;
+        if (rec.getReadNegativeStrandFlag()) {
+            clipPositionFrom = 0;
+            clipPositionTo = bases.length - clipFrom + 1;
+        } else {
+            clipPositionFrom = clipFrom - 1;
+            clipPositionTo = readLength;
+        }
+
+        String basesToKeepInTag = StringUtil.bytesToString(Arrays.copyOfRange(bases, clipPositionFrom, clipPositionTo));
+        String qualitiesToKeepInTag = SAMUtils.phredToFastq(Arrays.copyOfRange(baseQualities, clipPositionFrom, clipPositionTo));
+
+        if (rec.getReadNegativeStrandFlag()) {
+            // Ensures that bases are reverse complemented and base qualities are reversed
```

Comment:
"ensures that the qualities and bases in the tags are stored in their original order, as produced by the sequencer"